### PR TITLE
H-754, H-3333: Move `Filter` and `Subgraph` to store crate

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Strictly, they don't belong to the store but they are very related. This simplifies a lot of things. It's required to move them in order to move the store to a more accessible location.

## 🔍 What does this change?

- Split `QueryRecord` and `SubgraphRecord` (long overdue)
- Move `Filter`, `Subgraph`, and accommodating API to `hash-graph-store`

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: apps/hash-graph/libs/store/src/subgraph/record.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+use temporal_versioning::{ClosedTemporalBound, TemporalTagged, TimeAxis};
+
+use crate::subgraph::identifier::{
+    DataTypeVertexId, EntityTypeVertexId, EntityVertexId, PropertyTypeVertexId, VertexId,
+};
+
+/// A record persisted in the [`store`].
+///
+/// [`store`]: crate::store
+// TODO: Split trait into subgraph part and query part
```
Comment (indietyp) [inline apps/hash-graph/libs/store/src/subgraph/record.rs:14]: isn't this resolved in this very PR?

**Review by indietyp — APPROVED** [review]
minor non-blocking question, otherwise LGTM

File: apps/hash-graph/libs/store/src/subgraph/record.rs (hunk 1 above)
Comment (TimDiekmann) [reply apps/hash-graph/libs/store/src/subgraph/record.rs]: Yes, removed that 😅

**Review by indietyp — APPROVED** [review]

**Comment by TimDiekmann** [comment]
Sorry, I've overseen a private intra-doc link 🙃

**Pull request merged**